##### Import Library

In [18]:
import numpy as np
import pandas as pd

##### Load Dataset

In [19]:
titanic_data = pd.read_csv('train.csv')
titanic_data.head()

PassengerId  Survived  Pclass   
0            1         0       3  \
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp   
0                            Braund, Mr. Owen Harris    male  22.0      1  \
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

##### Memisahkan fitur - fitur / kolom yang penting dan yang tidak penting. Pada kasus ini, fitur yang tidak penting adalah PassengerId, Name, Ticket, Cabin  

In [20]:
titanic_data.drop(['PassengerId', 'Name', 'Ticket', 'Cabin'], axis=1, inplace=True)
titanic_data.head()


Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    male  22.0      1      0   7.2500        S
1         1       1  female  38.0      1      0  71.2833        C
2         1       3  female  26.0      0      0   7.9250        S
3         1       1  female  35.0      1      0  53.1000        S
4         0       3    male  35.0      0      0   8.0500        S

### Memahami Data 
1. Survived coloum menjelaskan apakah penumpang selamat dari tenggelamnya kapal titanic atau tidak. <b>0 bernilai tidak selamat, sedangkan 1 bernilai selamat</b>

2. Pclass adalah kelas untuk seberapa banyak penumpang melakukan perjalanan, apakah pertama kali, kedua kalinya, dst

3. Sex adalah jenis kelamin. Pria atau Wanita

4. Age adalah umur penumpang kapal titanic

5. SibSp and Parch adalah jumlah saudara atau orang tua yang ikut atau berada dalam kapal titanic.

6. Embarked adalah menjelaskan di mana penumpang menaiki kapal. C = Cherbourg, Q = Queenstown, S = Southampton.

##### Mengubah kolom Gender agar memiliki nilai kategori dalam bentuk numerik berdasarkan Markdown memahami data. 

In [21]:
from sklearn import preprocessing

le=preprocessing.LabelEncoder()
titanic_data['Sex'] = le.fit_transform(titanic_data['Sex'].astype(str))
titanic_data.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    1  22.0      1      0   7.2500        S
1         1       1    0  38.0      1      0  71.2833        C
2         1       3    0  26.0      0      0   7.9250        S
3         1       1    0  35.0      1      0  53.1000        S
4         0       3    1  35.0      0      0   8.0500        S

##### Kolom Embarked dilakukan proses One Hot Encoding, yaitu teknik untuk merepresentasikan sebuah data berbentuk categorical variable diubah ke dalam bentuk numerical value bernilai 0 dan 1

In [22]:
titanic_data = pd.get_dummies(titanic_data,columns=['Embarked'])
titanic_data.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked_C  Embarked_Q   
0         0       3    1  22.0      1      0   7.2500       False       False  \
1         1       1    0  38.0      1      0  71.2833        True       False   
2         1       3    0  26.0      0      0   7.9250       False       False   
3         1       1    0  35.0      1      0  53.1000       False       False   
4         0       3    1  35.0      0      0   8.0500       False       False   

   Embarked_S  
0        True  
1       False  
2        True  
3        True  
4        True

##### Melakukan pengecekan nilai data yang kosong / hilang

In [23]:
titanic_data[titanic_data.isnull().any(axis=1)]

Survived  Pclass  Sex  Age  SibSp  Parch     Fare  Embarked_C   
5           0       3    1  NaN      0      0   8.4583       False  \
17          1       2    1  NaN      0      0  13.0000       False   
19          1       3    0  NaN      0      0   7.2250        True   
26          0       3    1  NaN      0      0   7.2250        True   
28          1       3    0  NaN      0      0   7.8792       False   
..        ...     ...  ...  ...    ...    ...      ...         ...   
859         0       3    1  NaN      0      0   7.2292        True   
863         0       3    0  NaN      8      2  69.5500       False   
868         0       3    1  NaN      0      0   9.5000       False   
878         0       3    1  NaN      0      0   7.8958       False   
888         0       3    0  NaN      1      2  23.4500       False   

     Embarked_Q  Embarked_S  
5          True       False  
17        False        True  
19        False       False  
26        False       False  
28         True       False  
..          ...         ...  
859       False       False  
863       False        True  
868       False        True  
878       False        True  
888       False        True  

[177 rows x 10 columns]

In [24]:
#Drop data row yang kosong / hilang / tidak memiliki nilai
titanic_data = titanic_data.dropna()

##### Mean Shift Clustering

In [25]:
#Import MeanShift dari library sklearn.cluster
from sklearn.cluster import MeanShift
#Mengatur nilai bandwith hyperparameter ke 30 agar kernel yang dihasilkan pendek dan lebar
analyzer = MeanShift(bandwidth=30) 
#Terapkan bandwith = 30 ke dalam dataset
analyzer.fit(titanic_data)

MeanShift(bandwidth=30)

In [26]:
#Fungsi pembantu untuk membantu estimasi nilai bandwith yang baik berdasarkan dataset
"""from sklearn.cluster import estimate_bandwith
estimate_bandwith(titanic_data)""" 

'from sklearn.cluster import estimate_bandwith\nestimate_bandwith(titanic_data)'

In [27]:
labels = analyzer.labels_

In [28]:
np.unique(labels)

array([0, 1, 2, 3, 4], dtype=int64)

##### Bandwith 30 menghasilkan 5 klaster. Setiap titik melakukan ke masing - masing klaster

In [29]:
#Menambah kolom baru dalam dataset untuk menunjukan klaster data yang bergantungan
titanic_data['cluster_group'] = np.nan
data_length=len(titanic_data)
for i in range(data_length):
    titanic_data.iloc[i,titanic_data.columns.get_loc('cluster_group')] = labels[i]
titanic_data.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked_C  Embarked_Q   
0         0       3    1  22.0      1      0   7.2500       False       False  \
1         1       1    0  38.0      1      0  71.2833        True       False   
2         1       3    0  26.0      0      0   7.9250       False       False   
3         1       1    0  35.0      1      0  53.1000       False       False   
4         0       3    1  35.0      0      0   8.0500       False       False   

   Embarked_S  cluster_group  
0        True            0.0  
1       False            1.0  
2        True            0.0  
3        True            1.0  
4        True            0.0

##### EDA untuk keseluruhan Dataset

In [30]:
titanic_data.describe()

Survived      Pclass         Sex         Age       SibSp       Parch   
count  714.000000  714.000000  714.000000  714.000000  714.000000  714.000000  \
mean     0.406162    2.236695    0.634454   29.699118    0.512605    0.431373   
std      0.491460    0.838250    0.481921   14.526497    0.929783    0.853289   
min      0.000000    1.000000    0.000000    0.420000    0.000000    0.000000   
25%      0.000000    1.000000    0.000000   20.125000    0.000000    0.000000   
50%      0.000000    2.000000    1.000000   28.000000    0.000000    0.000000   
75%      1.000000    3.000000    1.000000   38.000000    1.000000    1.000000   
max      1.000000    3.000000    1.000000   80.000000    5.000000    6.000000   

             Fare  cluster_group  
count  714.000000     714.000000  
mean    34.694514       0.315126  
std     52.918930       0.690647  
min      0.000000       0.000000  
25%      8.050000       0.000000  
50%     15.741700       0.000000  
75%     33.375000       0.000000  
max    512.329200       4.000000

Dari EDA ini kia bisa melihat beberapa hal berikut : 

1) Total penumpang berjumlah 715

2) Dari 715 penumpang hanya 40 % nya saja yang selamat

3) PClass = 2.237 menunjukan banyak penumpang yang berada di kabin kelas bawah 

4) Rata - rata kolom Sex adalah 0.63 dimana 0 adalah wanita dan 1 adalah pria artinya, lebih banyak pria dibandingkan wanita

5) Biaya rata - rata penumpang sebesar $35

##### Melakukan EDA dengan Cluster

In [31]:
#Kelompokan penumpang berdasarkan klaster nya
titanic_cluster_data = titanic_data.groupby(['cluster_group']).mean()
#Hitung penumpang untuk masing - masing klaster
titanic_cluster_data['Counts'] = pd.Series(titanic_data.groupby(['cluster_group']).size())
titanic_cluster_data

Survived    Pclass       Sex        Age     SibSp     Parch   
cluster_group                                                                
0.0            0.336918  2.525090  0.679211  28.256720  0.439068  0.370968  \
1.0            0.611111  1.296296  0.527778  36.148148  0.814815  0.500000   
2.0            0.733333  1.000000  0.366667  32.430667  0.600000  0.866667   
3.0            0.733333  1.000000  0.266667  30.333333  1.000000  1.333333   
4.0            1.000000  1.000000  0.666667  35.333333  0.000000  0.333333   

                     Fare  Embarked_C  Embarked_Q  Embarked_S  Counts  
cluster_group                                                          
0.0             15.434139    0.121864    0.046595    0.831541     558  
1.0             65.622688    0.333333    0.018519    0.629630     108  
2.0            131.183883    0.500000    0.000000    0.500000      30  
3.0            239.991940    0.533333    0.000000    0.466667      15  
4.0            512.329200    1.000000    0.000000    0.000000       3

### Penjelasan 

Klaster ke 1 

- Terdapat 558 penumpang
- Persentase selamatnya adalah 33 % (sangat rendah), kebanyakan dari mereka tidak selamat
- Sebagian besar berasal dari kelas bawah, kelas 2 dan 3, dan sebagian besar berjenis kelamin laki - laki
- Biaya rata - rata yang dibayarkan adalah $15

Klaster ke 2 

- Terdapat 108 penumpang 
- Persentase selamatnya adalah 61 % (lebih dari separuhnya selamat)
- Sebagian besar berasal dari kelas 1 dan kelas 2
- Biaya rata - rata yang dibayarkan $65 

Klaster ke 3 

- Terdapat 30 penumpang 
- Persentase selamatnya adalah 73 % artinya, banyak dari mereka selamat 
- Sebagian besar berasal dari kelas 1 
- Biaya rata - rata yang dibayarkan $131 (Biaya tinggi)

Klaster ke 4 

- Terdapat 15 penumpang
- Persentase selamatnya 73 % artinya, banyak dari mereka selamat 
- Sebagian besar berasal dari kelas 1 
- Sebagian besar berjenis kelamin wanita 
- Biaya rata - rata yang dibayarkan $239 (Biaya tertinggi dibandingkan dengan rata - rata klaster pertama )

Klaster terakhir 
- Klaster ini dapat diabaikan karena hanya memiliki 3 datapoint, jadi tidak sangat signifikan, maka dari itu dapat diabaikan saja